pvHelpers/email/server_message_helpers.py

```diff
@@ -0,0 +1,133 @@
+        if not verifyServerMessage(user, message, raw_private_metadata,
+                                   fetchUser):
+            return False, None
+
+        decrypted_private_metadata = jloads(utf8Decode(raw_private_metadata))
+
+        bccs = []
+        if not crypto_url:
+            recipients = jloads(utf8Decode(user_encryption_key.unseal(
+                b64dec(message["wrapped_recipients"]))))
+
+            if message["wrapped_bccs"]:
+                bccs = jloads(utf8Decode(user_encryption_key.unseal(
+                    b64dec(message["wrapped_bccs"]))))
+        else:
+            # from postlord
+            recipients = user_encryption_key(
+                user.user_id,
+                key_version,
+                b64dec(message["wrapped_recipients"]),
+                crypto_url=crypto_url)
+            recipients = jloads(utf8Decode(recipients))
+
+            if message["wrapped_bccs"]:
+                bccs = user_encryption_key(user.user_id, key_version,
+                                           b64dec(message["wrapped_bccs"]),
+                                           crypto_url=crypto_url)
+                bccs = jloads(utf8Decode(bccs))
+
+        # verify the integrity of the wrapped_recipients against private metadata
+        if CaseInsensitiveSet(
+            map(lambda u: u["user_id"], recipients)) != CaseInsensitiveSet(
+                map(
+                    lambda u: u["user_id"], decrypted_private_metadata["ccs"] +
+                    decrypted_private_metadata["tos"])):
+            g_log.error(u"Server wrapped recipients don't match those of tos + ccs in private metadata")
+            return False, None
+
+        decrypted_private_metadata["bccs"] = bccs
+        return True, MergeDicts(message, {
+            "body": decrypted_private_metadata["body"],
+            "private_metadata": decrypted_private_metadata,
+            "timestamp": calendar.timegm(time.strptime(message["timestamp"], "%Y-%m-%dT%H:%M:%S")),
+            "attachments": decrypted_private_metadata["attachments"]
+        })
+    else:
+        decrypted_private_metadata = jloads(utf8Decode(mail_decrypt_key.decrypt(b64dec(message["private_metadata"]))))
+        if not verifyServerMessage(user, message, decrypted_private_metadata,
+                                   fetchUser):
+            return False, None
+
+        return True, MergeDicts(message, {
+            "body": MergeDicts(message["body"], {"snippet": utf8Decode(mail_decrypt_key.decrypt(b64dec(message["body"]["snippet"])))}),
+            "private_metadata": decrypted_private_metadata,
+            "timestamp": calendar.timegm(time.strptime(message["timestamp"], "%Y-%m-%dT%H:%M:%S")),
+            "attachments": map(lambda att: MergeDicts(att, {"name": utf8Decode(mail_decrypt_key.decrypt(b64dec(att["name"])))}), message["attachments"])
+        })
+
+
+def getWrappedKey(server_message):
```

Comment:
can we add ` @WrapException(KeyError, EmailException) ` to this so that consumers could catch issues with email structure w `EmailException`